app/src/main/kotlin/compiler/intermediate_form/SpillsHandlingAllocation.kt

```diff
@@ -1,20 +1,150 @@
+    data class Result(
         val allocatedRegisters: Map<Register, Register>,
         val linearProgram: List<Asmable>,
-        val spilledOffset: ULong
+        val spilledOffset: ULong,
     )
 
+    // Registers should be sorted by likeliness to be reserved for spilled
     fun allocateRegistersWithSpillsHandling(
         linearProgram: List<Asmable>,
         livenessGraphs: Liveness.LivenessGraphs,
-    ): Allocation.AllocationResult {
-        // Try allocating registers from H - R, and then handle spills with R (for |R| = 0, 1, 2, ...)
-        // + compute offset needed to allocate on stack for spilled registers
-        // + create linear program with definitions and usages inserted in appropriate places
-        return TODO()
+        orderedPhysicalRegisters: List<Register>,
+        allocator: Allocation
+    ): Result {
+        var reservedRegistersNumber = 0
+        while (true) {
+            val availableRegisters = orderedPhysicalRegisters.drop(reservedRegistersNumber)
+            val allocationResult = allocator.allocateRegisters(livenessGraphs, availableRegisters)
+            val spilledRegisters = allocationResult.spilledRegisters.toSet()
+
+            val instructionsToSpilled: Map<Int, SpilledInstruction> = linearProgram
+                .filterIsInstance<Instruction>()
+                .associate { linearProgram.indexOf(it) to toSpilledInstruction(it, spilledRegisters) }
+            val maxInstructionSpill = instructionsToSpilled.values.maxOfOrNull { it.numberOfSpilledRegisters } ?: 0
+
+            if (maxInstructionSpill <= reservedRegistersNumber) {
+
+                val spilledRegistersColoring: Map<Register, ULong> = colorSpilledRegisters(
+                    spilledRegisters,
+                    livenessGraphs,
+                    allocator,
+                )
+
+                val registerAllocation: Map<Register, Register> = allocateSpilledRegisters(
+                    orderedPhysicalRegisters.take(reservedRegistersNumber),
+                    allocationResult.allocatedRegisters,
+                    instructionsToSpilled.values.toList(),
+                )
+
+                val newLinearProgram: List<Asmable> = transformProgram(
+                    linearProgram,
+                    instructionsToSpilled,
+                    registerAllocation,
+                    spilledRegistersColoring,
+                )
+
+                return Result(
+                    registerAllocation,
+                    newLinearProgram,
+                    (spilledRegistersColoring.values.maxOrNull() ?: 0u) * memoryUnitSize,
+                )
+            }
+
+            reservedRegistersNumber ++
+        }
+    }
+
+    data class SpilledInstruction(
+        val instruction: Instruction,
+        val spilledUsedRegisters: Set<Register>,
+        val spilledDefinedRegisters: Set<Register>,
+        val numberOfSpilledRegisters: Int = (spilledUsedRegisters + spilledDefinedRegisters).size,
```

Comment:
Does this `+` work like a union of sets?

Reply:
yes